x-pack/plugins/security_solution/public/common/hooks/use_value_changed.ts

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import deepEqual from 'fast-deep-equal';
+import { useEffect, useRef } from 'react';
+
+export const useValueChanged = <T>(callback: (value: T) => void, nextValue: T) => {
+  const prevValue = useRef(nextValue);
+
+  useEffect(() => {
+    if (!deepEqual(prevValue.current, nextValue)) {
```

Comment:
ouch, a fast deepEqual is needed here? That always makes me think we have something else wrong somewhere?

I would maybe document the intent of this and when it should and shouldn't be used? Maybe add unit tests for it to make it clear. Seeing both the `useRef` and the `deepEqual` together already makes me feel like I would only use this hook as a last resort or weird component interaction and it can have a heavy potential cost?

It would be good to add docs on when _not_ to use it if it's easy to be a pattern a lot of people copy but isn't intended but for small select use cases?

We go to great lengths to ensure that if an array is mutated such as `selectedRuleIds` we create a new array reference along with the strings inside of it by using splats, concat, etc... It's really so we can pass `selectedRuleIds` into different dependencies of ReactJS which I do believe use the `Object.is` which is similar to the `===` reference compare and as fast as you can get.

The code where this is being used with `selectedRuleIds` actually has other spots where `selectedRuleIds` is being used in a react deps array which is going to be using `Object.is`. When I see the two being mixed I start to wonder as a maintainer if `selectedRuleIds` is being mutated with the same reference or not?

I think the `fast-deep-equal` is not something very good in general. It's easy to fix bugs quickly but the cost of it adds up fast and we begin to see mistakes. If you really need this and want it for others to use, I would consider possibly using a comparator function that can be passed in and that comparator function taking a `Object.is` function as the default and then using that where and when possible?

Fwiw, `fast-deep-equal` has a few bugs too, including where you must have a prototype:
https://github.com/epoberezkin/fast-deep-equal/issues/111

Reply:
A referential equality check (`Object.is` or `===`) wouldn't work in many cases, unfortunately. It has nothing to do with mutation, we do not mutate arrays, or at least I've never seen that. But there are places where a new array is created with the same values. A quick test shows that during the initial page load, the callback is called tens of times with a new empty array on every render.

```js
export const useValueChanged = <T>(callback: (value: T) => void, nextValue: T) => {
  const prevValue = useRef(nextValue);

  useEffect(() => {
    console.log({
      prev: prevValue.current,
      next: nextValue,
      deep: deepEqual(prevValue.current, nextValue),
      referential: prevValue.current === nextValue,
    });
    if (!deepEqual(prevValue.current, nextValue)) {
      prevValue.current = nextValue;
      callback(nextValue);
    }
  }, [callback, nextValue]);
};
```

Gives us the following picture:

<img width="985" alt="Screenshot 2021-06-04 at 20 53 16" src="https://user-images.githubusercontent.com/1938181/120849837-ec86c480-c576-11eb-8ef4-25d209ab4fc5.png">

Of course, we could try to eradicate every creation of a new empty array, but it will be too brittle. An unintentional `selectedRuleIds : []` or `selectedRuleIds = rules.map(({ id }) => id)` somewhere in the parent component would ruin performance.  Because wasted re-renders are much more expensive than the `deepEqual` comparison.

Reply:
>wasted re-renders are much more expensive than the deepEqual comparison.

I agree there, but I would look at things longer term and maybe what are we going to do if we end up with a lot of `deepEqual` or `deepEqual` everywhere and bad performance issues we then need to fix from there? It becomes painful when the whole application begins relying on `deepEqual` everywhere and someone has to spend a lot of time untangling it all vs. slowly improving our application to remove it and keep it from showing up again.

I would maybe take a small step back and play with the performance tooling to see if we are seeing bad effects of this already?

I typically don't want the ReactJS applications I work on to de-evolve into utilizing any form of `deepEqual` everywhere as the final solution, and there are strong warnings from ReactJS developers such as Dan Abramov:
https://twitter.com/dan_abramov/status/1104414707082055680

From Dan
```
There’s a reason React’s useEffect does *not* work this way. Slightly smiling face
```

```
Also “fast-deep-equal” is such a misleading library name. If it’s deep it can’t claim to be “fast” because you’re at the mercy of your data structure. Some seemingly innocent changes to it can make the comparison super slow.
```

about specifically `deepEqual`. With that said, "Have I worked w/ a lot of ReactJS applications that ended up mostly implementing deepEqual?" Yes, yes, I have. "Have I worked with deepEqual where it wasn't a performance concern?", Yes, yes I have. "Have I worked on applications without a lot of `React.memo()` since people knew when and where things were re-rendering and were thoughtful in the code reviews about it as well as the space issues?" Yes, I have as well.

Personally I avoid `fast-deep-equals` usage as like Dan I disagree with the name and feel that by definition it cannot ever really be fast. If I have to implement a deep equals I typically do it with strong code comments and cautions or just even a simple, "JSON.stringy()` as my compare or a custom compare function. The additional verbosity and ugliness is my hope that it scares others off from copying the pattern. :-)  

I would also consider this as already something we have accidentally inflicted upon ourselves and now we have a bigger problem as we are propagating and encouraging a fallacy with `deepEquals` as a good fix when in the long run it will probably have to be fixed and removed when we can do it.

I leave it up to you, but I would think about maybe if it's not the best long term how _could we_ encourage a better culture and environment and codebase that shows people how to do the "right thing" with regards to performance? 

Some notes here though that others shared with me:
----

Usually for the new arrays that are created which are `[]` most people fix those by implementing a stable ref in the codebase and then pushing down the `[]`.

Something like this within a ts file/module

```ts
export const stableRefArray = []
```

Quick searches show others trying to accomplish it:
https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/public/common/components/charts/barchart.tsx#L144

That, people already do, and should be encouraging as well as fixing when they see it.

The other tricky part is when people do a `.map`, `.filter`, etc... and that will always create a new shallow reference for an array, object, etc... when the fact is that maybe it's best to not do those or check for something before causing the new shallow reference to be created.

Reply:
I’ve added a comment to `useValueChanged`  and replaced `deepEqual`  with `JSON.stringify`. Completely remove `useValueChanged` we can after more thorough refactoring of the `RulesTables`: https://github.com/elastic/kibana/issues/92160#issuecomment-856634592